packages/devtools_app/lib/src/shared/console/eval/eval_service.dart

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:vm_service/vm_service.dart';
+
+import '../../../service/vm_service_wrapper.dart';
+import '../../object_tree.dart';
+import '../primitives/eval_history.dart';
+
+class AutocompleteCache {
+  final classes = <ClassRef, Class>{};
+
+  /// Cache of autocomplete matches for a library for code written within that
+  /// library.
+  ///
+  /// This cache includes autocompletes from all libraries imported and exported
+  /// by the library as well as all private autocompletes for the library.
+  final libraryMemberAndImportsAutocomplete =
+      <LibraryRef, Future<Set<String?>>>{};
+
+  /// Cache of autocomplete matches to show for a library when that library is
+  /// imported.
+  ///
+  /// This cache includes autocompletes from libraries exported by the library
+  /// but does not include autocompletes for libraries imported by this library.
+  final libraryMemberAutocomplete = <LibraryRef, Future<Set<String?>>>{};
+
+  void clear() {
+    classes.clear();
+    libraryMemberAndImportsAutocomplete.clear();
+    libraryMemberAutocomplete.clear();
+  }
+}
+
+class EvalService {
+  EvalService({
+    required this.isolateRef,
+    required this.variables,
+    required this.frameForEval,
+    required this.isPaused,
+    required this.service,
+  });
+
+  final IsolateRef? Function() isolateRef;
```

Comment:
It would seem more consistent to use ValueListenable for all of these rather than callbacks.

Reply:
ValueListenable looks too powerful for me. I want it to be clear that EvalService does not need to listen for the changes. It just needs up-to-date value.

Reply:
I don't quite buy this. The autocomplete cache is bogus if the isolate changed, for example. ValueListenable does not force you to always listen for changes. As appropriate you can always just use .value to get the current value. On the other end of the spectrum, VmServiceWrapper needs to just be a fixed instance because there is no reason anyone writing this code should worry it can changed.,

Reply:
My goal is clarity of contract for the method.
I want the constructor parameters to be smallest needed set of what the object needs to function. Of course, not at cost of convenience.  

About ValueListenable  vs function. I remember a story: when I passed ValueListenable, I got comments in code review like 'why this object need to listen to every change? it is performance overhead'. And then I needed to explain that the object just needs values.

So, it affects readability of code too.

Does it change something for you?

Reply:
About VmServiceWrapper: as I understand when reconnect happens, DebuggerController is not recreated, but it resets some variables. 
So, EvalService will not recreate, but just reset. 

Yes, it may be not the best implementation, but it is what we currently have. And it will take more refactoring to change.

Reply:
And, I did not understand you point about consistency. In Flutter framework there is ton of methods and constructors that take functions, not listeners.

Reply:
Each time an app connects you get a new DebuggerController. How it happens flows back to how the provider for the DebuggerController is initialized.
The private getter for the service in debuggerController is really just calling
`serviceManager.service!`
Would be clearer to just call that global from EvalService rather than making it look like it is injected in.

Reply:
In general, I think ValueListenable is almost always the right choice if something can actually change. If it can't change then pass in the actual value when initializing.

Reply:
Converted service to variable, isPaused and variables to listenable.